### ISSUE TITLE
Using roll_level other than 'none' results in error

### ISSUE THREAD

**Issue by jskrist** [sot/sparkles#59]
When I run the method:
`run_aca_review(**kwargs)`
with the input:
`'roll_level': 'none'`
Everything runs as expected and a report is generated for all the acar tables provided, however, when the 'roll_level' is anything other than 'none', I get the following error:

```
<class 'AttributeError'>: 'NoneType' object has no attribute 'fid_set'
 Traceback (most recent call last):
   File "<string>", line 1, in <module>
   File "<string>", line 22, in <module>
   File "[PythonPath]\lib\site-packages\sparkles-4.1-py3.6.egg\sparkles\core.py", line 156, in run_aca_review
     aca.check_catalog()
   File "[PythonPath]\lib\site-packages\sparkles-4.1-py3.6.egg\sparkles\core.py", line 653, in check_catalog
     self.check_acq_p2()
   File "[PythonPath]\lib\site-packages\sparkles-4.1-py3.6.egg\sparkles\core.py", line 771, in check_acq_p2
     P2 = -np.log10(self.acqs.calc_p_safe())
   File "[PythonPath]\lib\site-packages\proseco-4.4-py3.6.egg\proseco\acq.py", line 664, in calc_p_safe
     p_acqs = [acq['probs'].p_acqs(acq['halfw'], man_err) for acq in self]
   File "[PythonPath]\lib\site-packages\proseco-4.4-py3.6.egg\proseco\acq.py", line 664, in <listcomp>
     p_acqs = [acq['probs'].p_acqs(acq['halfw'], man_err) for acq in self]
   File "[PythonPath]\lib\site-packages\proseco-4.4-py3.6.egg\proseco\acq.py", line 1232, in p_acqs
     fid_set = self.acqs().fid_set
 Error executing python statement: 'exec(Path(MX_VAR_reviewScript).open().read())' - 'NoneType' object has no attribute 'fid_set'
```

the script being executed is as follows:

```
full_report_kw_dict = {
"load_name": load_name,
"acas": acas,
"make_html": True,
"outdir": outdir,
"roll_level": roll_level}

path = Path(outdir)
if path.exists():
  shutil.rmtree(path)

from sparkles.core import run_aca_review
run_aca_review(**full_report_kw_dict)
```

**Comment by taldcroft** [comment]
@jskrist - just quickly, are you using the latest master of both proseco and sparkles?  That exception is one that I spent a while squashing out of the code, but maybe I still missed something.

**Comment by jskrist** [comment]
@taldcroft - I updated to the latest from master in both proseco and sparkles this morning.

**Comment by taldcroft** [comment]
Are you able to reproduce the problem with this script?
```
from proseco import get_aca_catalog
from sparkles.core import run_aca_review, ACAReviewTable
from proseco.tests.test_common import mod_std_info

aca = get_aca_catalog(**mod_std_info(n_guide=3))
aca
acar = ACAReviewTable(aca)
run_aca_review(acas=[acar], load_name='load-name', make_html=True, outdir='test-outdir', roll_level='all')
```

**Comment by jskrist** [comment]
No, the error does not occur with that script.  Only a warning about not having a supplemental agasc file:

```
>>> from proseco import get_aca_catalog
>>> from sparkles.core import run_aca_review, ACAReviewTable
[pythonRoot]\lib\site-packages\proseco-4.4-py3.6.egg\proseco\characteristics.py:89: UserWarning: Unable to find C:\Users\jkristoff\Documents\MATLAB\FOT_Tools\Ska_data\data\agasc\agasc_supplement.h5, using limited bad star set
  warnings.warn(f'Unable to find {agasc_supplement_file}, using limited bad star set')
>>> from proseco.tests.test_common import mod_std_info
>>> aca = get_aca_catalog(**mod_std_info(n_guide=3))
>>> aca
<ACATable length=11>
 slot  idx      id    type  sz   p_acq  ...   yang     zang    dim   res  halfw
int64 int32   int64   str3 str3 float64 ... float64  float64  int32 int32 int64
----- ----- --------- ---- ---- ------- ... -------- -------- ----- ----- -----
    0     1         3  FID  8x8   0.000 ...    40.01 -1871.10     1     1    25
    1     2         4  FID  8x8   0.000 ...  2140.23   166.63     1     1    25
    2     3         5  FID  8x8   0.000 ... -1826.28   160.17     1     1    25
    3     4    139192  BOT  6x6   0.981 ...   587.27   802.49    20     1   160
    4     5 611190016  BOT  6x6   0.982 ...   175.44 -1297.92    20     1   160
    5     6 688521312  BOT  6x6   0.978 ...  -739.68 -1799.86    20     1   160
    6     7 688523960  ACQ  6x6   0.983 ...  -202.71 -1008.91    20     1   160
    7     8 611192384  ACQ  6x6   0.980 ...  1053.38  -275.16    20     1   160
    0     9 611189488  ACQ  6x6   0.976 ...  1536.83  -786.12    20     1   160
    1    10 688522008  ACQ  6x6   0.971 ...  -743.78 -2100.94    20     1   160
    2    11    134680  ACQ  6x6   0.972 ...  1314.00  1085.73    20     1   160
>>> acar = ACAReviewTable(aca)
>>> run_aca_review(acas=[acar], load_name='load-name', make_html=True, outdir='test-outdir', roll_level='all')
```

**Comment by taldcroft** [comment]
That version I gave you doesn't generate catalogs for the roll options, so maybe it wasn't a good case.  Please try this one, which actually has a new roll option catalog.
```
In [35]: kwargs = {'att': (160.9272490316051, 14.851572261604668, 99.996111473617802),
    ...:  'date': '2019:046:07:16:58.449',
    ...:  'detector': 'ACIS-S',
    ...:  'dither_acq': (7.9992, 7.9992),
    ...:  'dither_guide': (7.9992, 7.9992),
    ...:  'focus_offset': 0.0,
    ...:  'man_angle': 1.792525648258372,
    ...:  'n_acq': 8,
    ...:  'n_fid': 3,
    ...:  'n_guide': 5,
    ...:  'obsid': 21477,
    ...:  'sim_offset': 0.0,
    ...:  't_ccd_acq': -11.14616454993262,
    ...:  't_ccd_guide': -11.150381856818923}
    ...:  

In [36]: aca = get_aca_catalog(**kwargs)

In [37]: acar = aca.get_review_table()

In [38]: rm -rf junk

In [39]: run_aca_review(acas=[acar], outdir='junk', make_html=True, loud=True, load_name='junk', roll_level='all')
Processing obsid 21477
Writing output review file junk/index.html
```
Also, if there is any way you can print the `aca.meta['call_args']` for the catalog where this is failing, that should let me directly try to reproduce it locally.

**Comment by jskrist** [comment]
No errors running that code.  I will try to get the `aca.meta['call_args']` data for you from the catalog that is breaking.

**Comment by jskrist** [comment]
OK, so I can find an ACAReview Table that causes the error, and retrieve it's `meta['call_args']`, however, I cannot reproduce the error using those call args.  Furthermore, this happens:

```
>> print(acar == acas[3])
  [ True  True  True  True  True  True  True  True]
>> full_report_kw_dict["acas"] = [acas[3]]
>> run_aca_review(**full_report_kw_dict)
  Processing obsid 48470.0
   <class 'AttributeError'>: 'NoneType' object has no attribute 'fid_set' [rest of error...]
>> full_report_kw_dict["acas"] = [acar]
>> run_aca_review(**full_report_kw_dict)
   Processing obsid 48470.0
   Writing output review file [outdir]\index.html
```

Or in other words.  I have two ACAReview Tables that are supposedly equal, but when I use one, I get an error, and when I use the other, I don't.  I haven't looked into how the `==` operator is overloaded for this type, so there could be some aspects that are different, but I don't know what.

**Comment by jskrist** [comment]
I have found that if I do the following:

```
kw_acas = [item.meta["call_args"] for item in acas]
acas = [get_aca_catalog(**kwargs) for kwargs in kw_acas]
acars_clean = [aca.get_review_table() for aca in acas]
```
then I can get the `run_aca_review()` to run without errors, but it takes a really long time.

**Comment by taldcroft** [comment]
What's the code to create the ACAReviewTable objects in `acas`?

This issue is thorny and very deeply embedded in the structure.  First off, the `acar == acas[3]` equality is only testing for equality of the table row data, so that output is just saying that each of the 8 rows match (i.e. not that helpful).  The issue at hand as buried way below that level.  You should be able to see a difference if you look at:
```
acar.acqs.cand_acqs['probs'][0].acqs  # should show a weakref object with a "live" id
acas[3].acqs.cand_acqs['probs'][0].acqs  # should show a weakref object labeled as "dead"
```
Each `AcqProbs` object in the `'probs'` column in the candidate acquisition stars table needs to store a reference to its parent `acqs` table, and to avoid a reference cycle that Python doesn't seem able to garbage collect, it uses a weakref object.  But if you do something like make a copy of the table then the new version `probs` can be pointing to the original `acqs` instead of the copy, and if the original `acqs` gets deleted then the copy shows the exception you are seeing.

Someday in person I can explain the madness if you care, but for now it means that somehow your ACAReviewTable objects are corrupted by not having that weakref set.  It is supposed to happen here:
https://github.com/sot/sparkles/blob/master/sparkles/core.py#L325

**Comment by taldcroft** [comment]
So the question is whether anything is being done to the original ACATable objects before getting ACAReviewTable versions via `get_review_table()`.

The test code that I wrote for this might suggest something to you?  At least it states what should and should not work:
https://github.com/sot/sparkles/blob/86ecb2bc59c8b92cdbad90f19618c591538d42a7/sparkles/tests/test_review.py#L63

**Comment by taldcroft** [comment]
BTW you can put tests in the code like `acar.acqs.cand_acqs['probs'][0].acqs() is None` to see if/when the reference died.

**Comment by jskrist** [comment]
Thanks @taldcroft .  I am making a `deepcopy()` of the ACAReviewTable objects in order to store them for each OR/ER in the schedule.  I will check the status of the weakref you mentioned, but I suspect that this reference is being broken when I try to copy the object.

Basically, I am looping through each observation, and processing it and storing the kwargs used to create the ACATable along with copies of the ACATable and ACAReviewTable objects in a dictionary so I can get a "cached" version when nothing about the kwargs have changed, and mainly so I can create a list of the ACAReviewTables needed by the `run_aca_review()` function.

Is there a better way to store these objects so this reference (or any others) doesn't break?

**Comment by taldcroft** [comment]
You should just be able to do (effectively):
```
acas = {}
acars = {}
for obsid in obsids:
    aca = get_aca_catalog(**kwargs_for_obsid)
    acas[obsid] = aca
    acars[obsid] = aca.get_review_table()
```
No deepcopy required, and even though you are re-using `aca` the reference in the `acas` dict stays valid as the original.

BTW, the confusion between `ACATable` objects and `ACAReviewTable` objects is real, so I'm thinking about tidying up sparkles to be consistent everywhere in using `aca` and `acar` (and plurals `acas` and `acars`) respectively.  So this would be an API change in the `roll_options` dicts to use `acar` instead of `aca` and keyword args `acars` instead of `acas` (for the list of ACAReviewTable objects).

Thoughts?

**Comment by jskrist** [comment]
I welcome the consistency :)

**Comment by taldcroft** [comment]
OK will do.

Of course now I am trying to make the deepcopy() fail (as I did somehow when writing tests earlier), so I can fix it, and it doesn't fail!  Argh.  Anyway, if you take out the deepcopy() and this fixes this issue, then that would be at least progress forward.

**Comment by jskrist** [comment]
After removing the deepcopy() calls and storing the acars and acas as you suggested above, I still encountered the error about:
`'NoneType' object has no attribute 'fid_set'`
when running the `run_aca_review()` function.

I even added a section just above this call that loops through all the acars and attempts to update the acar if necessary:

```
acars = [acar for acar in acars_dict.values()]
# some of the weakrefs to the acaTables associated with the acaReviewTables
# get broken for unknown reasons.  Check the weakref here and fix if needed
for idx in range(len(acars)):
  acar = acars[idx]
  if acar.acqs.cand_acqs["probs"][0].acqs() is None:
    acars[idx] = acas_dict[acar.obsid].get_review_table()
```
but somehow, a couple lines later, when I use this list, one of the acars' weakref has died again.

**Comment by taldcroft** [comment]
Humm.  Can you try the latest master for both proseco and sparkles.  It might just work (after #278), but if not, master includes a new hammer method you can use to force the weakrefs.  This will work for ACATable or ACAReviewTable:
```
acar.acqs.set_cand_acqs_probs_weakref()
```
None of this should be necessary, so the ORviewer calling code must be doing something differently from what I'm expecting.  Are you going to be at OCC or Wayside tomorrow?  Maybe just sitting down together with the code is the way forward.

**Comment by jskrist** [comment]
I'll pull the latest and try again.  I am trying to create a reproduction script outside of MATLAB, and will send it your way if I succeed.  I planned to be working from home tomorrow (I am at the OCC today), but I could come in if this is still an issue at the end of the day.

**Comment by jskrist** [comment]
@taldcroft - after pulling the latest changes I am no longer getting this error; yay!

I am seeing some other weird behavior, but I'll create a separate issue for it once I am sure I am not just doing something weird on my end.

**Comment by jskrist** [comment]
@taldcroft - Ugh, this `'NoneType' object has no attribute 'fid_set'` bug is more persistent than I thought.  I am testing using a new, somewhat unusual schedule (ORViewer Workspace) and now I am getthing this issue again.  I will try to find the root cause, to see if it is due to a bad weakref, or something else...

**Comment by taldcroft** [comment]
:scream:
